public/app/features/alerting/unified/components/rule-editor/labels/LabelsField.tsx

```diff
@@ -112,29 +101,106 @@ const LabelsWithSuggestions: FC<{ dataSourceName: string }> = ({ dataSourceName
+  );
+  //------ Convert the labelsOpsKeys to the same format as the labelsByKeyFromExisingAlerts
+  const labelsByKeyOps = useMemo(() => {
+    return labelsOpsKeys.reduce((acc: Record<string, Set<string>>, label) => {
+      acc[label.name] = new Set();
+      return acc;
+    }, {});
+  }, [labelsOpsKeys]);
+
+  //------- Convert the keys from the ops labels to options for the dropdown
+  const keysFromGopsLabels = useMemo(() => {
+    return mapLabelsToOptions(Object.keys(labelsByKeyOps));
+  }, [labelsByKeyOps]);
+
+  //------- Convert the keys from the existing alerts to options for the dropdown
+  const keysFromExistingAlerts = useMemo(() => {
+    return mapLabelsToOptions(Object.keys(labelsByKeyFromExisingAlerts));
+  }, [labelsByKeyFromExisingAlerts]);
+
+  // create two groups of labels, one for ops and one for custom
+  const groupedOptions = [
+    {
+      label: 'From alerts',
+      options: keysFromExistingAlerts,
+    },
+    {
+      label: 'From system',
+      options: keysFromGopsLabels,
+    },
+  ];
+
+  const selectedKeyIsFromAlerts =
+    labelsByKeyFromExisingAlerts[selectedKey] !== undefined && labelsByKeyFromExisingAlerts[selectedKey]?.size > 0;
+  const valuesAlreadyFetched = !selectedKeyIsFromAlerts && labelsByKeyOps[selectedKey]?.size > 0;
+
+  // Only fetch the values for the selected key if it is from ops and the values are not already fetched (the selected key is not in the labelsByKeyOps object)
+  const {
+    currentData: valuesData,
+    isLoading: isLoadingValues = false,
+    error,
+  } = labelsApi.endpoints.getLabelValues.useQuery(
+    { key: selectedKey },
+    {
+      skip: !labelsPluginInstalled || !selectedKey || selectedKeyIsFromAlerts || valuesAlreadyFetched,
+    }
+  );
+
+  // these are the values for the selected key in case it is from ops
+  const valuesFromSelectedGopsKey = useMemo(() => {
+    // if it is from alerts, we need to fetch the values from the existing alerts
+    if (selectedKeyIsFromAlerts) {
+      return [];
+    }
+    // in case of a label from ops, we need to fetch the values from the plugin
+    // fetch values from ops only if there is no value for the key
+    const valuesForSelectedKey = labelsByKeyOps[selectedKey];
+    const valuesAlreadyFetched = valuesForSelectedKey?.size > 0;
+    if (valuesAlreadyFetched) {
+      return mapLabelsToOptions(valuesForSelectedKey);
+    }
+    if (!isLoadingValues && valuesData?.values?.length && !error) {
+      const values = valuesData?.values.map((value) => value.name);
+      labelsByKeyOps[selectedKey] = new Set(values);
+      return mapLabelsToOptions(values);
+    }
+    return [];
+  }, [selectedKeyIsFromAlerts, labelsByKeyOps, selectedKey, isLoadingValues, valuesData, error]);
 
   const getValuesForLabel = useCallback(
     (key: string) => {
-      return mapLabelsToOptions(labelsByKey[key]);
+      // values from existing alerts will take precedence over values from ops
```

Comment:
I decided to follow this approach, but happy to see your thoughts on this

Reply:
This seems like the most sensible approach 👍